components/raftstore-v2/src/operation/command/mod.rs

```diff
@@ -0,0 +1,165 @@
+use protobuf::Message;
+use raft::eraftpb::Entry;
+use raftstore::{
+    store::{
+        fsm::Proposal, local_metrics::RaftMetrics, metrics::*, msg::ErrorCallback, util,
+        WriteCallback,
+    },
+    Error, Result,
+};
+use slog::error;
+use tikv_util::{box_err, time::monotonic_raw_now};
+
+use crate::{
+    batch::StoreContext,
+    fsm::{PeerFsm, PeerFsmDelegate},
+    raft::Peer,
+    router::CmdResChannel,
+};
+
+mod write;
+
+pub use write::{RawWriteDecoder, RawWriteEncoder};
+
+impl<'a, EK: KvEngine, ER: RaftEngine, T> PeerFsmDelegate<'a, EK, ER, T> {
+    #[inline]
+    pub fn on_command(&mut self, req: RaftCmdRequest, ch: CmdResChannel) {
+        if !req.get_requests().is_empty() {
+            self.fsm
+                .peer_mut()
+                .on_write_command(self.store_ctx, req, ch)
+        } else if req.has_admin_request() {
+            // self.on_admin_request(req, ch)
+        } else if req.has_status_request() {
+            error!(self.fsm.logger(), "status command should be sent by Query");
+        }
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    #[inline]
+    fn validate_command(&self, req: &RaftCmdRequest, metrics: &mut RaftMetrics) -> Result<()> {
+        if let Err(e) = util::check_store_id(req, self.peer().get_store_id()) {
+            metrics.invalid_proposal.mismatch_store_id.inc();
+            return Err(e);
+        }
+        for r in req.get_requests() {
+            if let CmdType::Get | CmdType::Snap | CmdType::ReadIndex = r.get_cmd_type() {
+                return Err(box_err!("internal error: query can't be sent as command"));
+            }
+        }
+        if let Err(e) = util::check_peer_id(req, self.peer().get_id()) {
+            metrics.invalid_proposal.mismatch_peer_id.inc();
+            return Err(e);
+        }
+        if !self.is_leader() {
+            metrics.invalid_proposal.not_leader.inc();
+            return Err(Error::NotLeader(self.region_id(), self.leader()));
+        }
+        if let Err(e) = util::check_term(req, self.term()) {
+            metrics.invalid_proposal.stale_command.inc();
+            return Err(e);
+        }
+        if let Err(mut e) = util::check_region_epoch(req, self.region(), true) {
+            if let Error::EpochNotMatch(_, new_regions) = &mut e {
+                // TODO: query sibling regions.
+                metrics.invalid_proposal.epoch_not_match.inc();
+            }
+            return Err(e);
+        }
+        Ok(())
+    }
+
+    #[inline]
+    fn propose_command<T>(
+        &mut self,
+        ctx: &mut StoreContext<EK, ER, T>,
+        req: RaftCmdRequest,
+    ) -> Result<u64> {
+        let data = req.write_to_bytes().unwrap();
+        self.propose(ctx, data)
+    }
+
+    #[inline]
+    fn propose<T>(&mut self, ctx: &mut StoreContext<EK, ER, T>, data: Vec<u8>) -> Result<u64> {
+        ctx.raft_metrics.propose.normal.inc();
+        PEER_PROPOSE_LOG_SIZE_HISTOGRAM.observe(data.len() as f64);
```

Comment:
use local metrics?

Reply:
v1 use global metrics. Perhaps we can do that all together in other PR.